src/utils/pagedIterator.ts

```diff
@@ -0,0 +1,85 @@
+import { Issue as IssueRequest, Vault, Redeem, ActiveStakedRelayer } from "../interfaces/default";
+import { AugmentedQuery, PaginationOptions } from "@polkadot/api/types";
+import { StorageKey } from "@polkadot/types/primitive";
+import { Codec, Observable, CodecArg } from "@polkadot/types/types";
+import { ApiPromise } from "@polkadot/api";
+import { QueryableStorageEntry, ApiTypes } from "@polkadot/api/types";
+
+export type PolkadotCodecType = IssueRequest | Redeem | Vault | ActiveStakedRelayer;
+
+export async function* pagedIterator<T extends PolkadotCodecType>(
+    api: ApiPromise,
+    polkadotListings: AugmentedQuery<ApiTypes, (arg: CodecArg) => Observable<T>> & QueryableStorageEntry<ApiTypes>,
+    perPage: number
+): AsyncGenerator<T[]> {
+    let startKey: StorageKey = api.createType("Bytes", "") as StorageKey;
+    let issueRequests: [StorageKey, Codec][];
+    while (true) {
+        // a blank StorageKey is 0x in hex
+        if (startKey.toHex() == "0x") {
+            if (!isRemainingListNonempty(polkadotListings)) {
+                return undefined;
+            }
+            issueRequests = await listPaged(polkadotListings, perPage);
+        } else {
+            if (!isRemainingListNonempty(polkadotListings, startKey)) {
```

Comment:
Looking below, it appears that we get an error if there is nothing to return, so I would just add a `try/catch` and be done with it. 

```typescript
let startKey: StorageKey | undefined;
while (true) {
    try {
        const issueRequests = await listPaged(polkadotListings, perPage, startKey);
        if (issueRequests.length === 0) { // not sure if that's needed or if it always be an error in this case
            break;
        }
        startKey = issueRequests[issueRequests.length - 1][0];
        yield issueRequests.map((v) => v[1] as T);
    } catch (e) {
        break;
    }
}
```

Let me know if I'm missing something.

Reply:
That was my approach initially and it is indeed much simpler. I added a comment on line 70 justifying this solution:
```
// requesting ENTRIES using the last key as a cut-off point throws a substrate 
// error however, requesting KEYS with the last key only returns an empty array
// the latter method can be used to check if all entries were "consumed"
```.

So I need to check using the keys first